ENH: Further expansion of gufunc signature to allow broadcasting

mhvk opened the pull request:
Builds on #11175 to include allowing broadcasting in `gufuncs` (which now is surprisingly easy)>

With this, @mattharrigan's `all_equal` (#8528) would become truly useful, as it also allows things like `np.all_equal(ndarray, const, axis=?)`, i.e., one of the arguments can be broadcast over its core.

EDIT: has decent tests but does still need docs.

jaimefrio committed:
ENH: Add frozen dimensions to gufunc signatures

mattip committed:
ENH: convert reserved field to version, set new version to 1

So that we can keep track of expansions of the PyUFuncObject.

mattip committed:
ENH: test, document, implement flexible signature.

Goal is allow signatures like (m?,n),(n,p?)->(m?,p?) for matmul.

mhvk committed:
MAINT: Use flags instead of flexible_dims argument.

This allows future expansion and is arguably easier to follow.

Also adds tests, of the signature parsing, of the ignoring of
dimensions, and of use of the frozen dimensions via cross1d.

mhvk committed:
MAINT: Move dimension logic to its own routine; simplify.

It got very long and now matches logically the later call to get
core sizes.

mhvk committed:
MAINT: Remove n_dims_used, and reuse core_num_dims instead.

mhvk committed:
MAINT: Ensure signatures with frozen sizes written differently compare equal.

Was comparison by the string representation so that "03" and "3" would
be considered different dimensions.  That would be surprising for implementing
an inner loop...

mhvk committed:
ENH: Recognize |1 as can_broadcast in gufunc signature parser.

mhvk committed:
TST: add all_equal gufunc with can_broadcast signature.

But only test the non-broadcast version since that's all that will
work here.

mhvk committed:
ENH: Make broadcasting work for simple cases.

mhvk committed:
ENH: Allow core broadcasting also for missing dimensions.

eric-wieser commented:
To be clear, this only enables broadcasting when explicitly requested? I think doing it always would be harmful, and would let bugs happen silently.

mhvk commented:
@eric-wieser - yes, by adding `|1` to the dimension name or number -- clearly I should aim to add some docs! Example: for `all_equal`, the signature would become `(n|1),(n|1)->()` (as shown by the new example in `_umath.tests.src` - though the implementation there is a dumb one, unlike #8528).

shoyer commented on doc/source/reference/c-api.generalized-ufuncs.rst:
```diff
@@ -138,22 +141,35 @@ Notes:
 #. White spaces are ignored.
+#. An integer as a dimension name freezes that dimension to the value,
+#. The name can be suffixed with a question mark, this make the dimension a
+   core dimension only if it exists on the input or output, otherwise 1 is used.
```
Should be something like: `otherwise a dimensions of size 1 is inserted into the inputs and squeezed from the outputs.`

Also, we should explain exactly when optional dimensions are used. The rule from matmul is that optional dimensions are used whenever there are enough dimensions for them to be well defined, but matmul doesn't have optional core dimensions that appear on multiple arguments.

For example, how should a signature like `(n,m?),(n,m?)->()` be interpreted if the two inputs are 1d and 2d? There are two reasonable interpretations I can think of:
1. Using the signature `(n),(n)->()`, because not all arguments have enough dimensions.
2. Raising an error, because the behavior is potentially ambiguous.

mhvk replied:
O dear, what did I do - this discussion really belongs in @mattip's PR... Though in his PR I'm not exactly sure what the behaviour is...

Here, I do know: the moment any argument has too few dimensions, flexible dimensions are effectively removed altogether from the ufunc, i.e., it follows your option 1. I think this is most consistent with `matmul`, where, if one were to give an output, one would similarly expect that the dimension would just "not exist".

p.s. And apologies for not adding documentation on this beyond what @mattip had - do look at the tests - I have those for both `?` (flexible) and `|1` (broadcasting) - but still need to add some for numerical axes; so far, for those there are just tests that the signature gets parsed right.

(I had a separate PR where it was per argument, but it became clear to me that was just going to be a mess...)

mattip commented:
I would like to see the enhanced core signature functionality broken into a number of PRs. It seems this should be the last one, and needs some discussion on the mailing list.

mhvk committed:
DOC: document broadcastable and fixed-size core dimensions

mhvk commented:
Just pushed an update to the documentation explaining what the enhancements actually provide (given the rebase conflicts, the tests are currently meaningless, but off-line things pass).

mhvk commented:
@mattip - completely agree. I pushed this here mostly so that you could see how it worked with flags, and that it was easy to add further features.

Do you have specific suggestions on order? Frozen sizes is by far the easiest, but I'm happy to start with `?` as well, since we can then proceed with `matmul`. But perhaps for the ufunc struct, we can agree that we'll immediately add both `core_dim_sizes` and `core_dim_flags`?

njsmith commented:
The nice thing about `?` is that it has a clear use case (`matmul`) and has been discussed a fair amount on the list and elsewhere. For frozen dimensions and broadcasting core dimensions I'm much less comfortable that we understand the benefits and trade-offs of adding them.

njsmith commented:
Actually if I was going to suggest adding one thing to gufuncs besides `?`, it would be features to let us convert more of the popular numpy functions into gufuncs, like `sort`.

mhvk commented:
@njsmith - there is actually a very clear use case for broadcasting, the `all_equal` gufunc written by @mattharrigan (#8528), which makes *much* more sense with allowing broadcasting of one of the arguments (especially in combination with the `axis` keyword of #11018), as it can then truly be the short-circuit, memory-saving equivalent of `(a == b).all(axis=...)`

What is your other suggestion for an extension?

mhvk commented:
ps. For the frozen dimensions, I'd *really* like them - right now, I'm hacking the type-resolver to insist on dimensions of 3 in standards-of-astronomy code I'm wrapping in ufuncs; see https://github.com/mhvk/astropy/blob/88312819cdbf78605dfe5b9eb4c8245cae8bb887/astropy/_erfa/ufunc.c.templ#L587 (and see the code above for other hacks I have to do to make the machinery work...)

mhvk committed:
MAINT: Ensure correct order of inner strides for ignored dimensions.

Previously, when a core dimension at the end was missing, the strides
were not re-ordered accordingly, and the inner loop code needed to
check whether the corresponding dimension was zero. But if the strides
are passed on correctly, this is not needed at all, and the dimension
can just be unity (not indicating broadcasting this time).

mattip committed:
DOC: updates and fixes for gufunc descriptions.

mhvk committed:
DOC: Release notes for frozen dimensions.

mhvk committed:
DOC: Release note for flexible dimensions.

mhvk committed:
MAINT: factor out initialization of variable parts of the ufunc.

That is, parts like core_num_dims which can get changed during
execution, e.g., because keepdims is set or a dimension is flexible.

mhvk committed:
MAINT: change the name of the actual core_num_dims.

Now op_core_num_dims to make clear it is the number of core dimensions
an operand has, rather than that assumed by the ufunc.
